src/interfaces/kubernetes_interface.py

```diff
@@ -0,0 +1,263 @@
+                stdin=False,
+                stdout=True,
+                tty=False
+        )
+        # TODO: stream result is just a string, so there is no error code to check
+        # ideally, we use a method where we can check for an error code, otherwise we will
+        # need to check for errors in the string (meh)
+        #
+        # if result.exit_code != 0:
+        #     raise Exception(
+        #         f"Command failed with exit code {result.exit_code}: {result.output.decode('utf-8')}"
+        #     )
+        return result
+
+    def get_bitcoin_debug_log(self, container_name: str):
+        logs = self.client.read_namespaced_pod_log(
+            name=container_name,
+            namespace="default",
+        )
+        return logs
+
+    def get_bitcoin_cli(self, tank: Tank, method: str, params=None):
+        if params:
+            cmd = f"bitcoin-cli -regtest -rpcuser={tank.rpc_user} -rpcport={tank.rpc_port} -rpcpassword={tank.rpc_password} {method} {' '.join(map(str, params))}"
+        else:
+            cmd = f"bitcoin-cli -regtest -rpcuser={tank.rpc_user} -rpcport={tank.rpc_port} -rpcpassword={tank.rpc_password} {method}"
+        return self.exec_run(tank.container_name, cmd, user="bitcoin")
+
+    def get_messages(self, a_name: str, b_ipv4: str, bitcoin_network: str = "regtest", namespace: str = "default"):
+        subdir = "/" if bitcoin_network == "main" else f"{bitcoin_network}/"
+        dirs = self.exec_run(a_name, f"ls /home/bitcoin/.bitcoin/{subdir}message_capture", namespace)
+        dirs = dirs.splitlines()
+        messages = []
+
+        for dir_name in dirs:
+            if b_ipv4 in dir_name:
+                for file, outbound in [["msgs_recv.dat", False], ["msgs_sent.dat", True]]:
+                    # Fetch the file contents from the container
+                    file_path = f"/home/bitcoin/.bitcoin/{subdir}message_capture/{dir_name}/{file}"
+                    blob = self.exec_run(a_name, f"cat {file_path}", namespace)
+
+                    # Parse the blob
+                    json = parse_raw_messages(blob, outbound)
+                    messages = messages + json
+
+        messages.sort(key=lambda x: x["time"])
+        return messages
+
+    def logs_grep(self, pattern: str, network: str):
+        compiled_pattern = re.compile(pattern)
+
+        # Fetch the logs from the pod
+        log_stream = self.client.read_namespaced_pod_log(
+                name=self.logs_pod ,
+                namespace=self.namespace,
+                timestamps=True,
+                _preload_content=False,
+        )
+
+        matching_logs = []
+        for log_entry in log_stream:
+            log_entry_str = log_entry.decode('utf-8').strip()
+            if compiled_pattern.search(log_entry_str):
+                matching_logs.append(log_entry_str)
+
+        return '\n'.join(matching_logs)
+
+    def generate_deployment_file(self, warnet):
+        """
+        TODO: implement this
```

Comment:
Right now there is no intermediate file describing the network. Up just creates k8s resources directly using the k8s python client.

In these comments Josie has suggested that perhaps we should create resource files

https://github.com/bitcoin-dev-project/warnet/blob/1bf977cf2948bfbaaf2ca14f1fadecfb620bb204/src/interfaces/kubernetes_interface.py#L233-L235

Personally I don't really understand the benefit of creating files and then deploying those files. I can see that you could parse those files to do the "warnet_from_deployment" without asking k8s for what it has running but that doesn't seem much of a problem to me.

Is there any other benefit to an intermediate file that I'm not thinking of?

Reply:
I had not heard of using that strategy.

Reply:
another point for having the files: they could be written to a PV, which would allow us to shut the cluster down and then restart and re-deploy warnet? but perhaps there are other mechanisms for doing this in k8s that are more elegant?

Reply:
I can see some times when it would be useful. I kept waiting for a point where it was needed and I hadn't hit that yet so just kept it dynamic.